File: python/interpreter.py
import sys
import argparse
from typing import List

MEMORY_SIZE = 30000

def interpret(code: str) -> None:
    memory: List[int] = [0] * MEMORY_SIZE
    ptr = 0

    def find_matching_bracket(start: int, direction: int) -> int:
        depth = 0
        while True:
            if code[start] == '[':
                depth += direction
            elif code[start] == ']':
                depth -= direction
            if depth == 0:
                return start
            start += direction

    i = 0
    while i < len(code):
        match code[i]:
            case '>':
                ptr = (ptr + 1) % MEMORY_SIZE
            case '<':
                ptr = (ptr - 1) % MEMORY_SIZE
            case '+':
                memory[ptr] = (memory[ptr] + 1) % 256
            case '-':
                memory[ptr] = (memory[ptr] - 1) % 256
            case '.':
                print(chr(memory[ptr]), end='', flush=True)
            case ',':
                memory[ptr] = ord(sys.stdin.read(1))
            case '[':
                if memory[ptr] == 0:
                    i = find_matching_bracket(i, 1)
            case ']':
                if memory[ptr] != 0:
                    i = find_matching_bracket(i, -1)
            case _:
                pass

        i += 1


def main() -> None:
    parser = argparse.ArgumentParser(description="Brainfuck Interpreter")
    parser.add_argument('file', nargs='?', help="Brainfuck source file")
    args = parser.parse_args()

    if not args.file:
        print("Usage: python3 compiler.py <file>")
        sys.exit(1)

    if not args.file.endswith('.bf'):
        print("Error: Input file must have a .bf extension", file=sys.stderr)
        sys.exit(1)

    try:
        with open(args.file, 'r') as file:
            code = file.read()
    except IOError as e:
        print(f"Error opening file: {e}", file=sys.stderr)
        sys.exit(1)
    
    interpret(code)

if __name__ == "__main__":
    main()

File: lua/interpreter.lua
local MEMORY_SIZE = 30000

local function interpret(code)
    local memory = {}
    for i = 1, MEMORY_SIZE do memory[i] = 0 end
    local ptr = 1

    local function find_matching_bracket(start, direction)
        local depth = 0
        local index = start
        while true do
            if code:sub(index, index) == "[" then
                depth = depth + direction
            elseif code:sub(index, index) == "]" then
                depth = depth - direction
            end
            if depth == 0 then return index end
            index = index + direction
        end
    end

    local i = 1
    while i <= #code do
        local char = code:sub(i, i)
        if char == ">" then
            ptr = (ptr % MEMORY_SIZE) + 1
        elseif char == "<" then
            ptr = (ptr - 2) % MEMORY_SIZE + 1
        elseif char == "+" then
            memory[ptr] = (memory[ptr] + 1) % 256
        elseif char == "-" then
            memory[ptr] = (memory[ptr] - 1) % 256
        elseif char == "." then
            io.write(string.char(memory[ptr]))
        elseif char == "," then
            memory[ptr] = string.byte(io.read(1) or "\0")
        elseif char == "[" then
            if memory[ptr] == 0 then
                i = find_matching_bracket(i, 1)
            end
        elseif char == "]" then
            if memory[ptr] ~= 0 then
                i = find_matching_bracket(i, -1)
            end
        end
        i = i + 1
    end
end

local function main()
    if #arg == 0 or not arg[1]:match("%.bf$") then
        print("Usage: lua interpreter.lua <file.bf>")
        return
    end

    local file_name = arg[1]
    local file, err = io.open(file_name, "r")
    if not file then
        print("Error opening file:", err)
        return
    end

    local code = file:read("*a")
    file:close()
    interpret(code)
end

main()

